feat: support webpack magic comments

NozomuIkuta opened the pull request:
This PR is to support #71 in some way.

I just read through the code of this module once, so I may not be misunderstanding the implementation and making mistakes in approach. But, still, I will submit this PR with _WIP_ in the title, so that everyone can point them out and/or have a discussion on my implementation.

---

According to [Webpack's documentation](https://webpack.js.org/api/module-methods/#magic-comments), there are 6 kinds of magic comments supported by Webpack 4+, and they can be combined. 

- `webpackInclude`
- `webpackExclude`
- `webpackChunkName`
- `webpackMode`
- `webpackPrefetch`
- `webpackPreload`

In the context of component importing, I think it would be enough to support only `preferch` and `preload` (at least to resolve #71).

---

The author of the issue wants to set magic comment per component in the place where it's used.

```html
<LazyHeader /* webpackPreload: true */ />
```

However, it will be hassle to detect them in build pipeline. 
In addition, I guess it is likely that lazy components are wanted to be prefetched or preloaded, regardless of the places they are used.
So, IMO, it will be reasonable to configure it before components are handled by the loader (or before they are imported).

---

My approach is to add 2 properties to objects of `components` array module option: `prefetch` and `preload`.
This would be simplest in terms of code and API change. In this way, all the components under specific directories are configured same.

2nd possible approach will be to provide a way to configure magic comments per component. 
In this approach, API would get to be a little bit complicated.

3rd approach will be to statically add magic comments in `import` statement in the loader.
I know "no option" is best design sometimes, but I'm not sure it's now.

I do appreciate any feedback.

---

P.S.

This is my very first PR to Nuxt Component modules, after Nuxt Content. 👶 
I hope that my PR will be consequently merged and used by Nuxt Core.

---

TODO

- [x] Verify approach
- [x] Take review and fix code
- [ ] Add tests
- [x] Update documentation

pi0 commented:
Seems a nice improvement 👍  /cc @clarkdo Confirming if nuxt supports webpack hints for server-renderer manifest.

clarkdo commented:
I’ll do some testing and double check in nuxt for SSR

clarkdo commented on templates/components/plugin.js:
```diff
@@ -6,7 +6,13 @@ import { wrapFunctional } from './utils'
-  return `  ${c.pascalName.replace(/^Lazy/, '')}: () => import('../${relativeToBuild(c.filePath)}' /* webpackChunkName: "${c.chunkName}" */).then(c => wrapFunctional(${exp}))`
+  const magicComments = [
+    `webpackChunkName: "${c.chunkName}"`,
+    c.prefetch ? `webpackPrefetch: ${c.prefetch ? 'true' : 'false'}` : false,
```
`webpackPrefetch` and `webpackPreload` expected true or a number, so I think we can remove `false`.

NozomuIkuta replied:
Thank you for pointing out!
I made the parts more type-strict.

clarkdo reviewed (commented):
Just tested, it works ok.

![image](https://user-images.githubusercontent.com/4312154/118551748-f2e30700-b755-11eb-94c8-60b49e7c6039.png)
![image](https://user-images.githubusercontent.com/4312154/118551775-fc6c6f00-b755-11eb-83df-696d6fead782.png)
![image](https://user-images.githubusercontent.com/4312154/118551869-1a39d400-b756-11eb-8d71-ef924c949cae.png)

NozomuIkuta commented:
@pi0 @clarkdo 

I modified the code according to @clarkdo 's comment, and updated README.

It seems that plugin changes on template files doesn't have unit tests.

I will remove the `WIP` prefix from the PR title, but is there anything that I should do?

pi0 commented on README.md:
```diff
@@ -304,6 +304,31 @@ export default {
+- Type: `Boolean/Number`
+- Default: `false`
+
+These properties are used for Wepack's magic comments, learn more in [Wepack's official documentation](https://webpack.js.org/api/module-methods/#magic-comments).
```
It would be nice to mention it is effective only for `Lazy*` components. During development, we don't enable loader so it affects every component but in production, detected components are directly imported via `installComponents`

NozomuIkuta replied:
Thank you, I updated README by 6c33178aa9d03d82d482fa39bf03c86c905c17b8.

pi0 reviewed (approved):
Overall looks good to me 👍 Doing for another local try before merge.

NozomuIkuta commented:
Sorry for newbie question, but what is difference between templates: `index.js` and `plugin.js`?
Should I modify `index.js` too?

As far as I see, the file isn't invoked from anywhere...?

pi0 commented:
@NozomuIkuta index file is mainly for custom integrations like testing and not used directly. But would be nice if you can add it to that file too so we won't forget about if used in the future.

Mergin to avoid stalling. PR for improvement welcome and thanks <3
